Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java

```diff
@@ -1646,19 +1651,102 @@ private void checkCardCanBeCopied(Card card1) {
-            Ability ability1 = card1.getAbilities().get(i);
-            Ability ability2 = card2.getAbilities().get(i);
-            if (!ability1.getClass().equals(ability2.getClass())) {
-                fail(card1, "copy", " miss copy in ability " + ability1.getClass().getName());
+    private void compareClassRecursive(Object obj1, Object obj2, Card originalCard, String msg, int maxDepth) {
+        if (obj1 == null && obj2 == null) {
+            return;
+        } else if (obj1 == null || obj2 == null) {
+            fail(originalCard, "copy", "not same class for " + msg + "]");
+        } else if (obj1.getClass() != obj2.getClass()) {
+            fail(originalCard, "copy", "not same class for " + msg + "<" + obj1.getClass() + ">" + "]");
+        } else if (obj1 == obj2) { // for instances mostly
+            return;
+        } else {
+            // Only recurse so much.
+            if (maxDepth == 0) {
+                return;
             }
-            for (int j = 0; j < ability1.getEffects().size(); j++) {
-                Effect effect1 = ability1.getEffects().get(j);
-                Effect effect2 = ability2.getEffects().get(j);
-                if (!effect1.getClass().equals(effect2.getClass())) {
-                    fail(card1, "copy", "miss copy in effect " + effect1.getClass().getName());
+            // Only recurse on those objects
+            if (obj1 instanceof MageObject || obj1 instanceof Filter || obj1 instanceof Condition || obj1 instanceof Effect
+                    || obj1 instanceof Ability || obj1 instanceof Mana || obj1 instanceof Cost || obj1 instanceof DynamicValue
+                    || obj1 instanceof Choice || obj1 instanceof TargetPointer) {
+                //System.out.println(msg);
+                Class class1 = obj1.getClass();
+                Class class2 = obj2.getClass();
+                do {
+                    if (class1 == null && class2 == null) {
+                        return;
+                    }
+                    if (class1 == null || class2 == null) {
+                        fail(originalCard, "copy", "not same class for " + msg + "<" + obj1.getClass() + ">" + "]");
+                        return;
+                    }
+                    List<Field> ability2Fields = Arrays.stream(class2.getDeclaredFields()).collect(Collectors.toList());
+                    int fieldIndex = 0;
+                    for (Field field1 : class1.getDeclaredFields()) {
+                        Field field2 = ability2Fields.get(fieldIndex);
+                        field1.setAccessible(true);
+                        field2.setAccessible(true);
+                        try {
+                            Object value1 = field1.get(obj1);
+                            Object value2 = field2.get(obj2);
+
+                            boolean doFieldRecurse = true;
+                            if (class1 == CardImpl.class) {
+                                if (field1.getName() == "spellAbility") {
```

Comment:
Search by name is bad (it can be renamed any time), try to check it by class and values like that:
* class check: `field1.getName() == "spellAbility"` -> `field1.getType() instanceof SpellAbility.class`
* value check: `((CardImpl) obj1).getSpellAbility()` -> `(SpellAbility) field1.get(obj1)`

Reply:
I do agree on the class check, but the getSpellAbility needs to be called for the value to be set:
![image](https://github.com/magefree/mage/assets/34709007/7c636c70-6335-44dd-bb51-32960851c506)

Reply:
I did end up adding a check that the 5 special fields are existing.
If someone rename or delete one of those, there will be verify fails, with the comment next to it asking to rework the relevant code block.